Improvement/react navigation upgrade 5

andrepimenta opened the pull request:
<!--
Thanks for your contribution!

Please ensure that any applicable requirements below are satisfied before submitting this pull request. This will help ensure a quick and efficient review cycle.
-->

**Description**

Updates navigation packages to react-navigation 5:
```"@react-navigation/bottom-tabs": "^5.11.11",
    "@react-navigation/compat": "^5.3.15",
    "@react-navigation/drawer": "^5.12.5",
    "@react-navigation/native": "^5.9.4",
    "@react-navigation/stack": "^5.14.5",
    "react-native-screens": "^3.3.0",
```
Reference on how to upgrade from 4: https://reactnavigation.org/docs/upgrading-from-4.x

Thinks to look out for @ibrahimtaveras00 :
- Navigation to another screen not working
- Headers not looking the same, for example, missing the back button or with incorrect titles 
- Inputs not pre-filled or information missing when navigating to a new screen. Since we are passing information using screen parameters and the way this is done has changed, there is now a probability that this does not work as expected for some screens.
- Warnings and errors that are thrown that were not there before.
- In general anything that is not working. Due to some screens relying on parameters that are sent to them, and since the way this is done has changed, there is now a probability that some things may not work as expected.


**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

**Issue**

Resolves #2077

andrepimenta committed:
Move to new dependencies

andrepimenta committed:
move navigation.getparam to route.params

andrepimenta committed:
add navigationOptions

andrepimenta committed:
Fix dismiss and drawer working

andrepimenta committed:
Fix remaining route params

andrepimenta committed:
Fix switch navigator and fiat on ramp nav

andrepimenta committed:
Fix routename and pop

andrepimenta committed:
Remove screen props

andrepimenta committed:
Fix deeplink navigation

andrepimenta committed:
Fix navigation for inner screens

andrepimenta committed:
Ignore navigation warning

andrepimenta committed:
Fix get route

andrepimenta committed:
Fix showing protect wallet modal

andrepimenta committed:
Add route to proptypes

andrepimenta committed:
Add more missing proptypes

andrepimenta committed:
Update tests

andrepimenta committed:
Fix some more navigations

andrepimenta committed:
Merge branch 'develop' into improvement/react-navigation-upgrade-5

# Conflicts:
#	app/components/Nav/Main/MainNavigator.js
#	app/components/Nav/Main/__snapshots__/index.test.js.snap
#	app/components/UI/ApproveTransactionReview/index.js
#	app/components/UI/CollectibleModal/index.test.js
#	app/components/UI/DrawerView/index.js
#	app/components/UI/FiatOrders/PaymentMethodApplePay/index.js
#	app/components/UI/FiatOrders/PaymentMethodSelector/wyreApplePay.js
#	app/components/UI/OptinMetrics/index.js
#	app/components/Views/CollectibleView/__snapshots__/index.test.js.snap
#	app/components/Views/CollectibleView/index.js
#	app/components/Views/CreateWallet/index.js
#	app/components/Views/Entry/index.js
#	app/components/Views/ImportFromSeed/index.js
#	app/components/Views/LockScreen/index.js
#	app/components/Views/Login/index.js
#	app/components/Views/OnboardingCarousel/index.js
#	app/components/Views/SyncWithExtensionSuccess/index.js

andrepimenta committed:
Update collectibles navigatoin

andrepimenta committed:
Merge branch 'develop' into improvement/react-navigation-upgrade-5

wachunei committed:
Use useNavigation

wachunei committed:
Use optional chaining on `dangerouslyGetParent()`

wachunei commented on app/components/Nav/App/index.js:
```diff
@@ -22,121 +19,105 @@ import Main from '../Main';
+	</Drawer.Navigator>
 );
 
+// Is this necessary?
```
I have no idea about this 🤔

andrepimenta replied:
@estebanmino do you know?

wachunei committed:
Reorganize nav imports in App/index

wachunei commented on app/components/UI/DrawerView/index.js:
```diff
@@ -789,7 +803,7 @@ class DrawerView extends PureComponent {
 	onSecureWalletModalAction = () => {
 		this.setState({ showProtectWalletModal: false });
-		this.props.navigation.navigate(this.props.passwordSet ? 'AccountBackupStep1' : 'SetPasswordFlow');
+		this.props.navigation.navigate('SetPasswordFlow', this.props.passwordSet && { screen: 'AccountBackupStep1' });
```
```ts
this.props.passwordSet && { screen: 'AccountBackupStep1' }
```

^ if `this.props.passwordSet` is `false`, navigate will be called as `navigate('SetPasswordFlow', false)`, according to docs the second argument must be a params object, I would not rely on passing a boolean, it better be undefined, wdyt?

andrepimenta replied:
Fixed thanks!

wachunei commented on app/components/UI/NavbarBrowserTitle/index.js:
```diff
@@ -87,11 +83,19 @@ class NavbarBrowserTitle extends PureComponent {
 	};
 
 	onTitlePress = () => {
-		const showUrlModal = this.props.navigation.getParam('showUrlModal', () => null);
+		const showUrlModal =
+			this.props.route.params?.showUrlModal ??
+			function() {
+				'';
+			};
 		showUrlModal({ urlInput: this.props.url });
 	};
```
I've seen this pattern a few times, I propose we change it.

`??` operator returns right side value only when left side value is nullish (`null` or `undefined`), which is the same condition `?.` operator uses.

In this case I suggest we change this for:

```ts
this.props.route.params?.showUrlModal?.({ urlInput: this.props.url })
```
Instead of creating noop functions

rickycodes replied:
I agree with this :100:

andrepimenta replied:
Yeah I basically blindly replaced every instance of `navigation.getParam('ABC', DEFAULT)` with `route.params?.ABC ?? DEFAULT` as they mention in the documentation. See https://reactnavigation.org/docs/upgrading-from-4.x#no-more-getparam

I did this to be faster because I didn't know how much more changes I still had to do to make this work, and also to make sure everything behaved the same (say something expecting null instead of undefined for some reason).

But yeah I think this is a good step to do now since everything else is working! Just changed this, thanks!!

wachunei commented on app/components/UI/Notification/index.js:
```diff
@@ -41,11 +41,11 @@ function Notification({
 
 	const isInBrowserView = useMemo(() => {
-		const routes = navigation.state.routes;
+		const routes = navigation.dangerouslyGetState().routes;
```
since this is already a functional component we should use
`const routes = useNavigationState(state => state.routes)` at top level instead of `dangerouslyGetState()`

andrepimenta replied:
Do you wanna do this part?

wachunei commented on app/components/UI/WhatsNewModal/index.js:
```diff
@@ -150,17 +151,9 @@ const WhatsNewModal = props => {
 	useEffect(() => {
 		if (props.enabled && !!featuresToShow) {
-			const route = findRouteNameFromNavigatorState(props.navigation.state);
+			const route = findRouteNameFromNavigatorState(props.navigation.dangerouslyGetState());
```
We should refactor `findRouteNameFromNavigatorState` to receive `routes` directly.
And in places we can obtain `routes` by using
```ts
const routes = useNavigationState(state => state.routes);
```

andrepimenta replied:
Do you wanna do this part?

wachunei replied:
I'll take care of this today

wachunei commented on app/components/Views/SendFlow/SendTo/index.js:
```diff
@@ -266,7 +283,7 @@ class SendFlow extends PureComponent {
 		}
 		//Fills in to address and sets the transaction if coming from QR code scan
-		const targetAddress = navigation.getParam('txMeta', null)?.target_address;
+		const targetAddress = (route.params?.txMeta ?? null)?.target_address;
```
Refactor as `route.params?.txMeta?.target_address`

andrepimenta replied:
My related comment: https://github.com/MetaMask/metamask-mobile/pull/2731#discussion_r649053478

Fixed!!

wachunei commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -126,8 +130,8 @@ class NetworkSettings extends PureComponent {
-		const { navigation, frequentRpcList } = this.props;
-		const network = navigation.getParam('network', undefined);
+		const { route, frequentRpcList } = this.props;
+		const network = route.params?.network ?? undefined;
```
I've seen this a few times too: `value ?? undefined`, I think we should remove `?? undefined` unless that case is we want to explicitly set a `null` as `undefined` (`null ?? undefined`)

andrepimenta replied:
My related comment: #2731 (comment)

Fixed!!

wachunei commented on app/components/Views/SimpleWebview/index.js:
```diff
@@ -35,7 +39,7 @@ export default class SimpleWebview extends PureComponent {
 
 	render() {
-		const uri = this.props.navigation.getParam('url', null);
+		const uri = this.props.route.params?.url ?? null;
 		if (uri) {
```
This is a similar case, we should remove nullish coalescing operator from `value ?? null`, unless we wan to explicitly convert `undefined` to `null`

andrepimenta replied:
My related comment: #2731 (comment)

Fixed!!

andrepimenta committed:
Removing ?? null and ?? undefined

wachunei committed:
Update findRouteNameFromNavigatorState

wachunei committed:
Remove noop functions

andrepimenta committed:
Adding testing for navigation

andrepimenta committed:
Add more navigation tests

andrepimenta committed:
Add comment explaining the test files

wachunei committed:
Remove unnecessary boolean casting

wachunei committed:
Remove unnecessary variable

wachunei committed:
Remove unnecessary boolean casting

ibrahimtaveras00 commented:
**Issue 1:**

On onboarding wizard when attempting to explore the browser doesn't work; seen here = https://recordit.co/0e8vcow8rv (start at 37 sec mark) 

![Screen Shot 2021-06-23 at 3 57 47 PM](https://user-images.githubusercontent.com/41752922/123162520-b1bdd100-d43e-11eb-9472-02667f30ada8.png)

**Issue 2:** 

When tapping on "Learn more about gas fees" on the educate gas fee
![image](https://user-images.githubusercontent.com/41752922/123163508-c9498980-d43f-11eb-9909-2e1d25d44184.png)

I end up getting the following errors:

![Screen Shot 2021-06-23 at 4 13 48 PM](https://user-images.githubusercontent.com/41752922/123163539-d36b8800-d43f-11eb-8395-a214543d20fc.png)

![Screen Shot 2021-06-23 at 4 13 32 PM](https://user-images.githubusercontent.com/41752922/123163564-db2b2c80-d43f-11eb-8559-962e2e8260ff.png)

andrepimenta committed:
Merge branch 'improvement/react-navigation-upgrade-5' of https://github.com/MetaMask/metamask-mobile into improvement/react-navigation-upgrade-5

andrepimenta committed:
Fixes navigation issues

ibrahimtaveras00 reviewed (changes requested):
Above issues have been resolved!

Other issues:

**Issue 4:**

tap on a txn and then tap on the view on etherscan, then tap X and tap back and go back to wallet and you will see this error:

![Screen Shot 2021-06-24 at 3 01 20 PM](https://user-images.githubusercontent.com/41752922/123341661-59f59780-d51c-11eb-8b90-8f5a9e6f36ac.png)

seen here = https://recordit.co/nZpQ7uIgr0

**Issue 5:**

I'm not seeing the seed recovery phrase video in onboarding flow, nor on settings > Security & Privacy on iPhone devices

seen here = https://recordit.co/tItYEecCF0

![Screen Shot 2021-06-24 at 6 23 57 PM](https://user-images.githubusercontent.com/41752922/123342234-60d0da00-d51d-11eb-87b7-f3699746d0ca.png)

**Issue 6:**

If I tap learn more in the sign message warning box, I get the following exception 

![Screen Shot 2021-06-24 at 6 14 28 PM](https://user-images.githubusercontent.com/41752922/123342338-92e23c00-d51d-11eb-97a2-14f285d98b99.png)

seen here = https://recordit.co/Jqcv4Ekyvh

steps:
- go to https://metamask.github.io/test-dapp
- connect wallet
- scroll down and tap on on "Sign" under ETH Sign category 
- tap on the learn more link in the warning box

**Issue 7:** 

Remind me later on the secure your wallet flow not behaving as expected 

seen here = https://recordit.co/O6RtMbNIyT

steps:
- create a new wallet
- on the "Secure your wallet" page, tap on remind me later
- tap inside the check box and tap Skip
- agree or disagree to metametrics
- tap no thanks on the onboarding tutorial
- on the protect your wallet sticky tap on protect wallet (notice how the sticky doesn't go away) and covers the Start CTA from being tapped 
- next tap remind me later to get rid of the sticky 
- now tap on the Remind me later button above the (Not recommended) text 
- notice how you're stuck on this page and you can no longer go back to wallet

**Issue 8:**

I get the following exception when tapping on the "Continue to purchase ETH" button on the gas education flow

![Screen Shot 2021-06-24 at 6 47 38 PM](https://user-images.githubusercontent.com/41752922/123342943-ae9a1200-d51e-11eb-8b0c-c49d119992bc.png)

seen here = https://recordit.co/u7kfPQcH03

**Issue 9:**

The title to view seed phrase is different than current prod 

![image](https://user-images.githubusercontent.com/41752922/123343048-f9b42500-d51e-11eb-8d46-98c3bc8767bb.png)

**Issue 10:**

The title to view private key is different than current prod 

![Screen Shot 2021-06-24 at 7 08 37 PM](https://user-images.githubusercontent.com/41752922/123343358-9bd40d00-d51f-11eb-8a8b-591e0e455371.png)

**Issue 11:**

On android, if I tap the device back button on metametrics screen it doesn't prevent me from going back nor show the alert box anymore compared to prod

<img src="https://user-images.githubusercontent.com/41752922/123344089-1e110100-d521-11eb-9a57-65fe4a1d025e.png" width=50% height=50%>

**Issue 12:**

When attempting to open a deeplink, I am taken to the browser, however, the URL does not load and I am just on the browser homepage.

seen here = http://recordit.co/wtTJ5pivh1

To reproduce, you can open this page on chrome and tap this link:
https://metamask.app.link/dapp/metamask.github.io/test-dapp/ 

**Issue 13:**

If I import a private key by scanning the QR directly from wallet view, I get the following error:

![Screen Shot 2021-06-24 at 7 46 55 PM](https://user-images.githubusercontent.com/41752922/123345829-f2901580-d524-11eb-8f74-93a98b51ee49.png)

steps:
- start on wallet view
- tap on QR scanner on top right
- scan a QR code for a private key

**Issue 14:**

I cannot scroll on the NFT modal on Android devices 

![image](https://user-images.githubusercontent.com/41752922/123346431-44856b00-d526-11eb-8486-150f76094383.png)

**Issue 15:**

When I relaunch Android, and log in, I notice the following on the bottom for a quick second, but then goes away:

![image](https://user-images.githubusercontent.com/41752922/123350495-1e14ff00-d529-11eb-9276-e925e0f60b74.png)

**Issue 16:**

The padding on the network and browser URL is incorrect on Android devices

![image](https://user-images.githubusercontent.com/41752922/123351995-6255ce80-d52c-11eb-8880-9d6055bf92e1.png)

andrepimenta committed:
Fix navigation issues

ibrahimtaveras00 commented:
**Issue 17:**

Reset wallet flow not working as expected.  on iOS, it seemed like it imported an account even though I didn't select anything, while on Android, I couldn't even get the modal to pop up to type "delete" 

seen here = http://recordit.co/esIXuDY7Ot

**Issue 18:**

Sync with extension navigation issues:
on iOS it immediately takes me to the login page after successfully
on Android it works, but opens up the QR scanner again when it shouldn’t

andrepimenta committed:
Merge branch 'develop' into improvement/react-navigation-upgrade-5

andrepimenta committed:
Fix QR scanner

ibrahimtaveras00 commented:
Issues 1 - 6 = Resolved 👍🏽 
**Issue 7** = still happens https://recordit.co/MvR0ygtB8Y
Issues 8 - 13 = Resolved 👍🏽 
**Issue 14** = still happens
Issues 15 - 18 = Resolved 👍🏽

sethkfman committed:
fixed enroll navigation reset issue

andrepimenta committed:
Fix going to wallet view

sethkfman committed:
updated test cases

andrepimenta committed:
Fix simple notification

ibrahimtaveras00 commented:
Latest Outstanding Issues:

- **Issue 14**
I cannot scroll on the NFT modal on Android devices

![image](https://user-images.githubusercontent.com/41752922/123873483-8a667880-d904-11eb-847b-00484a8c316f.png)

- **Issue 21**
delete/reset wallet flow no longer keeps user on "Wallet setup page" and doesn't show delete wallet notification seen here = http://recordit.co/RTogvGfObY

all other issues have been resolved 👍🏽

andrepimenta committed:
Fix going to onboarding

sethkfman committed:
added snapshots to test

sethkfman committed:
merged react native branch

sethkfman committed:
fixed what new modal test

sethkfman committed:
added picker to package.json

sethkfman committed:
yarn clean

sethkfman committed:
yarn update command

sethkfman committed:
snapshot update

sethkfman committed:
Merge branch 'improvement/react-navigation-upgrade-5' into improvement/react-navigation-upgrade-5-issue-19

sethkfman committed:
removed the dive method on the wrapper

sethkfman committed:
snapshot update

sethkfman committed:
update yarn unit test

sethkfman committed:
commented out cache in build step

Cal-L committed:
Bug/fix nft gesture experience (#2862)

* Fix collections modal

* Make eslint happy

* Clean up modal code
TODO: Converting components into hooks will save us from re-renders and from doing things like waiting to focus on inputs

* Remove in-house safe area from reusable modal for now

* Fix scroll interaction in collectible modal action sheet

* Use color from overlay

* Add test snap files

* Update snapshots and make tests pass

sethkfman committed:
uncommented cache for bitrise

ibrahimtaveras00 reviewed (approved):
All fixes look good, QA Passed 👍🏽
